### PR TITLE
Add zmon built in command for getting data from s3 buckets

### PR DESCRIPTION
**Pull request by lorenzhawkes**
Add a new s3_bucket() command that allows the user to fetch data from s3 buckets. 
The contents of the s3 bucket can be retrieved as is, or as JSON.

### CODE REVIEW

File: zmon_worker_monitor/builtins/plugins/s3_bucket.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        self.__client.download_fileobj(bucket_name, key, data)
+        result = data.getvalue()
+        data.close()
+        return result
+
+    def get_json_object(self, bucket_name, key):
+        data = StringIO.StringIO()
+        self.__client.download_fileobj(bucket_name, key, data)        
+        data.seek(0)
+        result = json.load(data)
```
Comment (hjacobs) [inline zmon_worker_monitor/builtins/plugins/s3_bucket.py]: always use `safe_load`!

**Comment by hjacobs** [comment]
I would rename the wrapper to `S3Wrapper` and only have `get_object` with another helper object to have a simpler way of invocation and to match the existing `requests` library style:

```
s3().get_object('my-bucket', 'foo/bar').text()
s3().get_object('my-bucket', 'foo/bar').json()

File: zmon_worker_monitor/builtins/plugins/s3_bucket.py (hunk 1 above)
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/s3_bucket.py]: I believe ``safe_load`` is in yaml and not json. But ``try...finally`` would be safer to call ``data.close``

File: zmon_worker_monitor/builtins/plugins/s3_bucket.worker_plugin [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[Core]
+Name = s3_bucket
+Module = s3_bucket
+
+[Documentation]
+Version = 0.1
+Website = https://github.com/zalando/zmon-worker
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/s3_bucket.worker_plugin]: `zalando-zmon/zmon-worker`

**Comment by lorenzhawkes** [comment]
:+1:

File: zmon_worker_monitor/builtins/plugins/s3.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import logging
+import requests
+import cStringIO
+
+from zmon_worker_monitor.adapters.ifunctionfactory_plugin import IFunctionFactoryPlugin, propartial
+
+logging.getLogger('botocore').setLevel(logging.WARN)
+
+
+class S3BucketWrapperFactory(IFunctionFactoryPlugin):
```
Comment (hjacobs) [inline zmon_worker_monitor/builtins/plugins/s3.py]: rename to `S3WrapperFactory`?

File: zmon_worker_monitor/builtins/plugins/s3.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    def get_object(self, bucket_name, key):
+        data = cStringIO.StringIO()
+        self.__client.download_fileobj(bucket_name, key, data)
+        result = data.getvalue()
+        data.close()
+        return S3Object(result)
+
+
+class S3Object(object):
```
Comment (hjacobs) [inline zmon_worker_monitor/builtins/plugins/s3.py]: rename to `S3Wrapper`?

**Comment by mohabusama** [comment]
Just couple of concerns:

1. What will happen with extremely large files? In the worker we define a hard limit on result size, but with s3 buckets we could load really large files before size limit assertion.
2. I imagine some would need to check if file exists or just its metadata (e.g size) without the need to download the object, so won't it be more convenient (+ efficient) to add ``.size()`` and ``.exists()``?

**Comment by lorenzhawkes** [comment]
I've added the metadata methods for an S3Object and to get them from the S3Bucket without having to fetch the entrire S3Object first of all.

I've updated the unit tests so they behave correctly under the error conditions I've observed when testing locally.

**Comment by hjacobs** [comment]
@mohabusama can you review & approve if you think it's ok?

Also important: documentation!

**Comment by lorenzhawkes** [comment]
Added docs for everything

**Comment by hjacobs** [comment]
@lorenzhawkes thanks. Sorry, I meant user docs on https://docs.zmon.io/: https://github.com/zalando-zmon/zmon-docs

File: zmon_worker_monitor/builtins/plugins/s3.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        return propartial(S3Bucket, region=factory_ctx.get('entity').get('region', None))
+
+
+def get_region():
+    r = requests.get('http://169.254.169.254/latest/dynamic/instance-identity/document', timeout=3)
+    return r.json()['region']
+
+
+class S3Bucket(object):
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/s3.py]: Please change to ``S3Wrapper`` for consistency with other plugins.

**Comment by mohabusama** [comment]
Just a general remark on the interface, a rather more consistent one could be:

```python
s3('my-bucket', 'my-key').size()
s3('my-bucket', 'my-key').exists()
s3('my-bucket', 'my-key').json()
s3('my-bucket', 'my-key').text()
```
so in a way, we would abstract ``get_object`` and ``get_object_metadata`` from the user.

However, we can go with current implementation and depend on proper documentation on https://docs.zmon.io

@hjacobs, @lorenzhawkes what do you think?

**Comment by hjacobs** [comment]
@mohabusama I think the current behavior is fine --- it's not consistent with HttpWrapper, but it makes sense here as we might want to have a call like `list_buckets` in the future.
I just kind of dislike the extra `get_object_metadata` (but I don't have a better suggestion now):

```
s3().get_object('my-bucket', 'my-key').json()
s3().list_buckets() # suggestion, maybe "ls" with pattern is better

**Comment by mohabusama** [comment]
👍

**Comment by hjacobs** [comment]
:+1:

**Pull request merged**